File: packages/client/src/lib/profileCss.ts


const MAX_LEN = 100_000;

const EXTERNAL_URL = /url\(\s*['"]?(?!data:)/i;
const DANGEROUS_TOKEN = /(expression\(|javascript:|-moz-binding|behavior\s*:|@import|@charset)/i;


const LAYER_NAME = /^[\w-]+(\.[\w-]+)*$/;

const STYLE_CLOSE = /<\/style/gi;


function stripStyleClose(css: string): string {
  let out = css;
  while (STYLE_CLOSE.test(out)) {
    STYLE_CLOSE.lastIndex = 0;
    out = out.replace(STYLE_CLOSE, "");
  }
  STYLE_CLOSE.lastIndex = 0;
  return out;
}


type GroupingRule = CSSRule & {
  cssRules: CSSRuleList;
  containerName?: string;
  containerQuery?: string;
  name?: string;
};


function isRule(name: string, rule: CSSRule): rule is GroupingRule {
  const ctor = (globalThis as Record<string, unknown>)[name];
  return typeof ctor === "function" && rule instanceof (ctor as new () => CSSRule);
}


function splitSelectorList(selector: string): string[] {
  const parts: string[] = [];
  let depth = 0;
  let current = "";
  for (const ch of selector) {
    if (ch === "(") depth++;
    else if (ch === ")") depth = Math.max(0, depth - 1);
    if (ch === "," && depth === 0) {
      parts.push(current);
      current = "";
    } else {
      current += ch;
    }
  }
  if (current.trim()) parts.push(current);
  return parts;
}


function scopeSelector(selector: string, scope: string): string {
  return splitSelectorList(selector)
    .map((s) => s.trim())
    .filter(Boolean)
    .map((s) => `${scope} ${s}`)
    .join(", ");
}

/** Rebuild a declaration block, dropping anything unsafe. */
function filterDeclarations(style: CSSStyleDeclaration): string {
  const out: string[] = [];
  for (let i = 0; i < style.length; i++) {
    const prop = style.item(i);
    const value = style.getPropertyValue(prop);
    if (!value) continue;
    if (DANGEROUS_TOKEN.test(value) || EXTERNAL_URL.test(value)) continue;
    if (prop === "position" && /\b(fixed|sticky)\b/i.test(value)) continue;
    const priority = style.getPropertyPriority(prop) ? " !important" : "";
    out.push(`${prop}: ${value}${priority}`);
  }
  return out.join("; ");
}

function processRules(rules: CSSRuleList, scope: string): string {
  let css = "";
  for (const rule of Array.from(rules)) {
    if (rule instanceof CSSStyleRule) {
      const selector = scopeSelector(rule.selectorText, scope);
      const decls = filterDeclarations(rule.style);
      if (selector && decls) css += `${selector} { ${decls} }\n`;
    } else if (rule instanceof CSSMediaRule) {
      const inner = processRules(rule.cssRules, scope);
      if (inner) css += `@media ${rule.media.mediaText} { ${inner} }\n`;
    } else if (rule instanceof CSSSupportsRule) {
      const inner = processRules(rule.cssRules, scope);
      if (inner) css += `@supports ${rule.conditionText} { ${inner} }\n`;
    } else if (isRule("CSSContainerRule", rule)) {
      // Container queries let a card react to its own width - the profile popup
      // and the settings preview render it at different sizes.
      const inner = processRules(rule.cssRules, scope);
      const name = LAYER_NAME.test(rule.containerName ?? "") ? `${rule.containerName} ` : "";
      if (inner) css += `@container ${name}${rule.containerQuery} { ${inner} }\n`;
    } else if (isRule("CSSStartingStyleRule", rule)) {
      // Entry animations without keyframes. Carries no prelude, so nothing to sanitize.
      const inner = processRules(rule.cssRules, scope);
      if (inner) css += `@starting-style { ${inner} }\n`;
    } else if (isRule("CSSLayerBlockRule", rule)) {
      // Cascade layers are purely local ordering. The name is an ident that can
      // encode braces via \XX escapes, so an unrecognised one becomes anonymous
      // rather than being spliced back into the sheet.
      const inner = processRules(rule.cssRules, scope);
      const name = LAYER_NAME.test(rule.name ?? "") ? ` ${rule.name}` : "";
      if (inner) css += `@layer${name} { ${inner} }\n`;
    } else if (rule instanceof CSSKeyframesRule) {
      // Keyframes are keyed by name and can only affect elements that reference
      // them (which are all scoped). Rebuild frames through the declaration
      // filter so a keyframe can't smuggle an external url().
      let frames = "";
      for (const kf of Array.from(rule.cssRules)) {
        if (kf instanceof CSSKeyframeRule) {
          const decls = filterDeclarations(kf.style);
          if (decls) frames += `${kf.keyText} { ${decls} }\n`;
        }
      }
      // `rule.name` is the *decoded* ident, and idents can encode `{`/`}` via
      // \XX escapes - re-escape it or the name splices raw rules past every
      // filter above (unscoped selectors, external url(), position:fixed).
      if (frames) css += `@keyframes ${CSS.escape(rule.name)} { ${frames} }\n`;
    }
    // Everything else (@import, @font-face, @page, …) is intentionally dropped.
  }
  return css;
}

/**
 * Returns scoped, safe CSS ready to drop into a <style> tag, or "" if the input
 * is empty or unparseable. `scopeClass` must be the class on the card container.
 */
export function sanitizeProfileCss(css: string | null | undefined, scopeClass: string): string {
  if (!css || !css.trim()) return "";
  try {
    const doc = document.implementation.createHTMLDocument("");
    const style = doc.createElement("style");
    style.textContent = css.slice(0, MAX_LEN);
    doc.head.appendChild(style);
    const sheet = style.sheet;
    if (!sheet) return "";
    return stripStyleClose(processRules(sheet.cssRules, `.${scopeClass}`));
  } catch {
    return "";
  }
}

File: android/app/src/main/java/lt/oranges/orangchat/util/ProfileCss.kt
package lt.oranges.orangchat.util

const val PROFILE_CSS_MAX_LEN = 100_000

private const val MAX_NESTING = 8

private val EXTERNAL_URL = Regex("""url\(\s*['"]?(?!data:)""", RegexOption.IGNORE_CASE)
private val DANGEROUS_TOKEN = Regex(
    """(expression\(|javascript:|-moz-binding|behavior\s*:|@import|@charset)""",
    RegexOption.IGNORE_CASE,
)
private val PROPERTY_NAME = Regex("""^-{0,2}[a-zA-Z][a-zA-Z0-9-]*$""")
private val LAYER_NAME = Regex("""^[\w-]+(\.[\w-]+)*$""")
private val CONTAINER_QUERY = Regex("""^[\w\s.,()<>=:%+*/-]+$""")
private val COMMENT = Regex("""/\*.*?\*/""", RegexOption.DOT_MATCHES_ALL)

private val STYLE_CLOSE = Regex("""</style""", RegexOption.IGNORE_CASE)

private fun stripStyleClose(css: String): String {
    var out = css
    while (STYLE_CLOSE.containsMatchIn(out)) out = STYLE_CLOSE.replace(out, "")
    return out
}

fun sanitizeProfileCss(css: String?, scopeSelector: String = ".oc-profile-card"): String {
    if (css.isNullOrBlank()) return ""
    val source = COMMENT.replace(css.take(PROFILE_CSS_MAX_LEN), " ")
    val processed = runCatching { processRules(source, scopeSelector, 0) }.getOrDefault("")
    return stripStyleClose(processed)
}

private fun processRules(input: String, scope: String, depth: Int): String {
    if (depth > MAX_NESTING) return ""
    val out = StringBuilder()
    var i = 0
    while (i < input.length) {
        val (terminator, at) = nextTopLevel(input, i) ?: break
        if (terminator == ';') {
            i = at + 1
            continue
        }
        val open = at
        val close = matchingBrace(input, open)
        if (close < 0) break
        val prelude = input.substring(i, open).substringAfterLast('}').trim()
        val body = input.substring(open + 1, close)
        i = close + 1

        if (prelude.startsWith("@")) {
            val name = prelude.takeWhile { !it.isWhitespace() }.lowercase()
            val params = prelude.drop(name.length).trim()
            if (DANGEROUS_TOKEN.containsMatchIn(params)) continue
            when {
                name == "@media" || name == "@supports" -> {
                    val inner = processRules(body, scope, depth + 1)
                    if (inner.isNotBlank()) out.append("$name $params {\n").append(inner).append("}\n")
                }
                name == "@container" && CONTAINER_QUERY.matches(params) -> {
                    val inner = processRules(body, scope, depth + 1)
                    if (inner.isNotBlank()) out.append("$name $params {\n").append(inner).append("}\n")
                }
                name == "@starting-style" && params.isBlank() -> {
                    val inner = processRules(body, scope, depth + 1)
                    if (inner.isNotBlank()) out.append("$name {\n").append(inner).append("}\n")
                }
                name == "@layer" && (params.isBlank() || LAYER_NAME.matches(params)) -> {
                    val inner = processRules(body, scope, depth + 1)
                    if (inner.isNotBlank()) {
                        out.append(if (params.isBlank()) "$name {\n" else "$name $params {\n")
                            .append(inner).append("}\n")
                    }
                }
                name.endsWith("keyframes") -> {
                    val frames = processKeyframes(body)
                    if (frames.isNotBlank() && params.isNotBlank()) {
                        out.append("$name $params {\n").append(frames).append("}\n")
                    }
                }
                else -> Unit
            }
        } else {
            val selector = scopeSelectorList(prelude, scope)
            val declarations = filterDeclarations(body)
            if (selector.isNotBlank() && declarations.isNotBlank()) {
                out.append("$selector { $declarations }\n")
            }
        }
    }
    return out.toString()
}

private fun processKeyframes(body: String): String {
    val out = StringBuilder()
    var i = 0
    while (i < body.length) {
        val open = body.indexOf('{', i)
        if (open < 0) break
        val close = matchingBrace(body, open)
        if (close < 0) break
        val keyText = body.substring(i, open).substringAfterLast('}').trim()
        val declarations = filterDeclarations(body.substring(open + 1, close))
        i = close + 1
        if (keyText.isNotBlank() && declarations.isNotBlank()) {
            out.append("$keyText { $declarations }\n")
        }
    }
    return out.toString()
}

private fun nextTopLevel(input: String, from: Int): Pair<Char, Int>? {
    var parens = 0
    var quote: Char? = null
    for (i in from until input.length) {
        val ch = input[i]
        if (quote != null) {
            if (ch == quote && input.getOrNull(i - 1) != '\\') quote = null
            continue
        }
        when (ch) {
            '"', '\'' -> quote = ch
            '(' -> parens++
            ')' -> parens = maxOf(0, parens - 1)
            '{', ';' -> if (parens == 0) return ch to i
        }
    }
    return null
}

private fun matchingBrace(input: String, open: Int): Int {
    var depth = 0
    var quote: Char? = null
    for (i in open until input.length) {
        val ch = input[i]
        if (quote != null) {
            if (ch == quote && input.getOrNull(i - 1) != '\\') quote = null
            continue
        }
        when (ch) {
            '"', '\'' -> quote = ch
            '{' -> depth++
            '}' -> {
                depth--
                if (depth == 0) return i
            }
        }
    }
    return -1
}

private fun splitTopLevel(input: String, separator: Char): List<String> {
    val parts = mutableListOf<String>()
    val current = StringBuilder()
    var depth = 0
    for (ch in input) {
        when (ch) {
            '(' -> depth++
            ')' -> depth = maxOf(0, depth - 1)
        }
        if (ch == separator && depth == 0) {
            parts.add(current.toString())
            current.clear()
        } else {
            current.append(ch)
        }
    }
    if (current.isNotBlank()) parts.add(current.toString())
    return parts
}

private fun scopeSelectorList(selector: String, scope: String): String {
    if (DANGEROUS_TOKEN.containsMatchIn(selector)) return ""
    if (selector.contains('<')) return ""
    return splitTopLevel(selector, ',')
        .map { it.trim() }
        .filter { it.isNotEmpty() }
        .joinToString(", ") { "$scope $it" }
}

private fun filterDeclarations(body: String): String {
    val kept = mutableListOf<String>()
    for (declaration in splitTopLevel(body, ';')) {
        val colon = declaration.indexOf(':')
        if (colon <= 0) continue
        val property = declaration.substring(0, colon).trim().lowercase()
        var value = declaration.substring(colon + 1).trim()
        if (!PROPERTY_NAME.matches(property) || value.isEmpty()) continue

        var important = false
        val bang = value.indexOf('!')
        if (bang >= 0) {
            important = value.substring(bang).trim().equals("!important", ignoreCase = true)
            value = value.substring(0, bang).trim()
        }
        if (value.isEmpty()) continue
        if (DANGEROUS_TOKEN.containsMatchIn(value) || EXTERNAL_URL.containsMatchIn(value)) continue
        if (property == "position" && Regex("""\b(fixed|sticky)\b""", RegexOption.IGNORE_CASE).containsMatchIn(value)) continue

        kept.add(if (important) "$property: $value !important" else "$property: $value")
    }
    return kept.joinToString("; ")
}
